src/game/logic/object/object.cpp

```diff
@@ -2655,51 +2984,468 @@ void Object::Enter_Group(AIGroup *group)
+                }
+
+                float range = Get_Template()->Get_Shroud_Reveal_To_All_Range();
+
+                if (range > 0.0f && !Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION)
+                    && (!Get_Status(OBJECT_STATUS_STEALTHED) || Get_Status(OBJECT_STATUS_DETECTED)
+                        || Get_Status(OBJECT_STATUS_DISGUISED))) {
+                    Coord3D pos = *Get_Position();
+                    unsigned short players =
+                        g_thePlayerList->Get_Players_With_Relationship(Get_Controlling_Player()->Get_Player_Mask(),
+                            PLAYER_RELATIONSHIP_FLAGS_ENEMIES | PLAYER_RELATIONSHIP_FLAGS_NEUTRALS);
+                    g_thePartitionManager->Do_Shroud_Reveal(pos.x, pos.y, range, players);
+                    m_allLookSighting->m_where = pos;
+                    m_allLookSighting->m_playerIndex = players;
+                    m_allLookSighting->m_radius = range;
+                }
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is looking, but hasn't unlooked the last one.");
+    }
+}
+
+void Object::Unlook()
+{
+    if (!m_friendlyLookSighting->Is_Invalid()) {
+        g_thePartitionManager->Queue_Undo_Shroud_Reveal(m_friendlyLookSighting->m_where.x,
+            m_friendlyLookSighting->m_where.y,
+            m_friendlyLookSighting->m_radius,
+            m_friendlyLookSighting->m_playerIndex);
+        m_friendlyLookSighting->Reset();
+
+        if (!m_allLookSighting->Is_Invalid()) {
+            g_thePartitionManager->Queue_Undo_Shroud_Reveal(m_allLookSighting->m_where.x,
+                m_allLookSighting->m_where.y,
+                m_allLookSighting->m_radius,
+                m_allLookSighting->m_playerIndex);
+            m_allLookSighting->Reset();
+        }
+    }
+}
+
+void Object::Shroud()
+{
+    if (m_shroudSighting->Is_Invalid()) {
+        if (Get_Controlling_Player() != nullptr) {
+            if (!Get_Status_Bits().Test(OBJECT_STATUS_UNDER_CONSTRUCTION) && !Is_Effectively_Dead()
+                && Get_Shroud_Range() > 0.0f) {
+                unsigned short index;
+
+                for (int i = g_thePlayerList->Get_Player_Count() - 1; i >= 0; i--) {
+                    Player *player = g_thePlayerList->Get_Nth_Player(i);
+
+                    if (Get_Controlling_Player()->Get_Relationship(player->Get_Default_Team()) == ALLIES) {
+                        index |= player->Get_Player_Mask();
+                    }
+                }
+
+                Coord3D pos = *Get_Position();
+                g_thePartitionManager->Do_Shroud_Cover(pos.x, pos.y, Get_Shroud_Range(), index);
+                m_shroudSighting->m_where = pos;
+                m_shroudSighting->m_playerIndex = index;
+                m_shroudSighting->m_radius = Get_Shroud_Range();
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is shrouding, but hasn't unshrouded the last one.");
```

Comment:
Can add object name into this message. Otherwise this assert does not tell much.